src/uri.jl

```diff
@@ -153,46 +153,24 @@ function Base.isvalid(uri::URI)
 end
 
 # RFC3986 Unreserved Characters (and '~' Unsafe per RFC1738).
-@inline shouldencode(c) = !((c >= UInt8('A') && c <= UInt8('Z'))
-                         || (c >= UInt8('a') && c <= UInt8('z'))
-                         || (c >= UInt8('0') && c <= UInt8('9'))
-                         || c == UInt8('-')
-                         || c == UInt8('.')
-                         || c == UInt8('_'))
-hexstring(x) = string('%', uppercase(hex(x,2)))
+@inline isunsafe(c::Char) = !(c in "-._") && (!isascii(c) || !isalnum(c))
```

Comment:
No `in`s please; they're way too slow vs. explicit `x == y || ...` comparisons

Reply:
Only ~10% slower in context:

```julia
julia> @inline shouldencode(c) = !((c >= UInt8('A') && c <= UInt8('Z'))
                                 || (c >= UInt8('a') && c <= UInt8('z'))
                                 || (c >= UInt8('0') && c <= UInt8('9'))
                                 || c == UInt8('-')
                                 || c == UInt8('.')
                                 || c == UInt8('_'))

julia> @inline isunsafe(c::Char) = !(c in "-._") && (!isascii(c) || !isalnum(c))

julia> utf8_chars(str::AbstractString) = (Char(c) for c in Vector{UInt8}(str))

julia> escape(c::Char) = string('%', uppercase(hex(c,2)))

julia> newescape(str::AbstractString, check::Function=isunsafe) =
           join(check(c) ? escape(c) : c for c in utf8_chars(str))

julia> oldescape(str::AbstractString, check::Function=shouldencode) =
           join(check(Int(c)) ? escape(c) : c for c in utf8_chars(str))

julia> d = String(Vector{UInt8}(rand(0:255,1000000)))
1000000-element Array{UInt8,1}

julia> bnew = @benchmarkable newescape(d)

julia> bold = @benchmarkable oldescape(d)

julia> run(bold)
BenchmarkTools.Trial:
  memory estimate:  281.08 MiB
  allocs estimate:  5966566
  --------------
  minimum time:     225.803 ms (11.35% GC)
  median time:      234.066 ms (11.58% GC)
  mean time:        248.353 ms (17.21% GC)
  maximum time:     319.342 ms (36.01% GC)
  --------------
  samples:          21
  evals/sample:     1

julia> run(bnew)
BenchmarkTools.Trial:
  memory estimate:  281.08 MiB
  allocs estimate:  5966566
  --------------
  minimum time:     246.922 ms (10.48% GC)
  median time:      256.928 ms (11.49% GC)
  mean time:        268.947 ms (15.56% GC)
  maximum time:     342.289 ms (33.90% GC)
  --------------
  samples:          19
  evals/sample:     1

julia> d = Vector{UInt8}(rand(0:255,10))
10-element Array{UInt8,1}

julia> run(bold)
BenchmarkTools.Trial:
  memory estimate:  3.13 KiB
  allocs estimate:  69
  --------------
  minimum time:     2.474 μs (0.00% GC)
  median time:      2.602 μs (0.00% GC)
  mean time:        2.932 μs (8.06% GC)
  maximum time:     2.427 ms (97.39% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> run(bnew)
BenchmarkTools.Trial:
  memory estimate:  3.13 KiB
  allocs estimate:  69
  --------------
  minimum time:     2.701 μs (0.00% GC)
  median time:      2.825 μs (0.00% GC)
  mean time:        3.198 μs (7.53% GC)
  maximum time:     2.478 ms (97.23% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> d = Vector{UInt8}(rand(0:255,1000))
1000-element Array{UInt8,1}

julia> run(bold)
BenchmarkTools.Trial:
  memory estimate:  291.33 KiB
  allocs estimate:  6068
  --------------
  minimum time:     198.116 μs (0.00% GC)
  median time:      208.255 μs (0.00% GC)
  mean time:        247.330 μs (11.94% GC)
  maximum time:     4.300 ms (93.01% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> run(bnew)
BenchmarkTools.Trial:
  memory estimate:  291.33 KiB
  allocs estimate:  6068
  --------------
  minimum time:     220.771 μs (0.00% GC)
  median time:      230.028 μs (0.00% GC)
  mean time:        278.296 μs (11.06% GC)
  maximum time:     3.848 ms (89.57% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

Reply:
A little tweak to make the `in` list `const` reduces the slowdown to ~5%:
```julia
const safe = Vector{UInt8}("-._")
@inline isunsafe(c) = !(UInt8(c) in safe) && (!isascii(c) || !isalnum(c))
```

Reply:
Custom `isalnum()` to avoid ccalls into utf8 libraries reduces the slowdown to ~2%:
```julia
@inline isascii(c::UInt8) = c < UInt8(0x80)

@inline isalnum(c::UInt8) = ((c >= UInt8('A') && c <= UInt8('Z'))
                          || (c >= UInt8('a') && c <= UInt8('z'))
                          || (c >= UInt8('0') && c <= UInt8('9')))

const safe = Vector{UInt8}("-._")

@inline isunsafe(c::UInt8) = !(c in safe) && (!isascii(c) || !isalnum(c))
```

Reply:
However, when I change the benchmark to escape a more real-world example of escaping a query string Dict() the slowdown is around 3-4% even without these tweaks.
So, in summary, the "No ins please; they're way too slow" does not seem to apply in this use case.
If the benchmark was to construct a URI, including escaping a query string, the slowdown would be on the order of 1-2%. If the benchmark was `GET`-ing a URI, even from localhost, the slowdown would be well and truely in the noise.